Userland/Libraries/LibCompress/Lzma.cpp

```diff
@@ -0,0 +1,604 @@
+        //  "IsMatch" and "IsRep0Long" arrays."
+        u16 position_state = m_total_decoded_bytes & ((1 << m_options.position_bits) - 1);
+        u16 state2 = (m_state << maximum_number_of_position_bits) + position_state;
+
+        auto update_state_after_literal = [&] {
+            if (m_state < 4)
+                m_state = 0;
+            else if (m_state < 10)
+                m_state = m_state - 3;
+            else
+                m_state = m_state - 6;
+        };
+
+        auto update_state_after_match = [&] {
+            if (m_state < 7)
+                m_state = 7;
+            else
+                m_state = 10;
+        };
+
+        auto update_state_after_rep = [&] {
+            if (m_state < 7)
+                m_state = 8;
+            else
+                m_state = 11;
+        };
+
+        auto update_state_after_short_rep = [&] {
+            if (m_state < 7)
+                m_state = 9;
+            else
+                m_state = 11;
+        };
+
+        auto copy_match_to_buffer = [&](u16 real_length) -> ErrorOr<void> {
+            VERIFY(!m_leftover_match_length.has_value());
+
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() < m_total_decoded_bytes + real_length)
+                return Error::from_string_literal("Tried to copy match beyond expected uncompressed file size");
+
+            while (real_length > 0) {
+                if (m_output_buffer.empty_space() == 0) {
+                    m_leftover_match_length = real_length;
+                    break;
+                }
+
+                u8 byte;
+                auto read_bytes = TRY(m_output_buffer.read_with_seekback({ &byte, sizeof(byte) }, m_rep0 + 1));
+                VERIFY(read_bytes.size() == sizeof(byte));
+
+                auto written_bytes = m_output_buffer.write({ &byte, sizeof(byte) });
+                VERIFY(written_bytes == sizeof(byte));
+                m_total_decoded_bytes++;
+
+                real_length--;
+            }
+
+            return {};
+        };
+
+        // If we have a leftover part of a repeating match, we should finish that first.
+        if (m_leftover_match_length.has_value()) {
+            TRY(copy_match_to_buffer(m_leftover_match_length.release_value()));
+            continue;
+        }
+
+        // "The decoder uses the following code flow scheme to select exact
+        //  type of LITERAL or MATCH:
+        //
+        //  IsMatch[state2] decode
+        //   0 - the Literal"
+        if (TRY(decode_bit_with_probability(m_is_match_probabilities[state2])) == 0) {
+            // "At first the LZMA decoder must check that it doesn't exceed
+            //  specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding literal symbol");
```

Comment:
we already check this above, so this is unrechable. It would only be reachable if we did another iteration of the loop to check for the EOF marker, which we don't.

Reply:
(I would keep the spec text, but add a comment explaining it's not unecessary, assuming you agree with me of course :slightly_smiling_face:)